### PR TITLE
[Feature] Realistic Explosions

### PR DESCRIPTION
**Pull request by DeityLink**
- Added a new realistic method for processing explosions that takes walls and some other obstacles into account (closed Airlocks and Blast Doors)
- Admins can at any time toggle the explosion method (between the old one and the new one) from the Secret panel (requires +SERVER)
- Reinforced Floors are now actually bomb resistant, and less likely to break apart from explosions (they don't affect the spread of the explosion at all though, nor the damage to atoms standing on top of it)
- Blast Doors are now less likely to be destroyed by explosion (previously they were as fragile as any other doors)
- Unsimulated/Shuttle floors now get covered in soot when hit by explosions
- Added a verb to the debug panel that lets you visualize in colours the impacts of an explosion (can be set to use either explosion methods, the colours disappear after 10 seconds)

update
newcops

### CODE REVIEW

**Comment by ComicIronic** [comment]
This is a strong amount of hardcode.

File: code/game/objects/explosion.dm [hunk 1]
```diff
@@ -95,6 +95,27 @@
 		for (var/turf/T in trange(max_range, epicenter))
 			var/dist = cheap_pythag(T.x - x0, T.y - y0)
 
+			if(explosion_newmethod)	//realistic explosions that take obstacles into account
+				var/turf/Trajectory = T
```
Comment (RemieRichards) [inline code/game/objects/explosion.dm]: no point defining this, T is already /turf

File: code/game/objects/explosion.dm [hunk 2]
```diff
@@ -95,6 +95,27 @@
 		for (var/turf/T in trange(max_range, epicenter))
 			var/dist = cheap_pythag(T.x - x0, T.y - y0)
 
+			if(explosion_newmethod)	//realistic explosions that take obstacles into account
+				var/turf/Trajectory = T
+				while(Trajectory != epicenter)
+					Trajectory = get_step_towards(Trajectory,epicenter)
+					if(istype(Trajectory,/turf/simulated/wall) || istype(Trajectory,/turf/simulated/shuttle/wall))
```
Comment (RemieRichards) [inline code/game/objects/explosion.dm]: You can probably fix all this snowflake by giving atoms an explosion_resistance variable, and distance simply goes up by how resistant the atom is (for the case of Poddoors below going up by 2, etc.)

File: code/game/objects/explosion.dm (hunk 1 above)
Comment (DeityLink) [inline code/game/objects/explosion.dm]: In the code as it is right now yes, but I was sleep deprived when I coded that, I'm rewriting it so it's less "hardcoded" as Comic is pointing out. Turfs and Doors will now have an explosion_block var, so I need to keep Trajectory defined as turf.

Comment (RemieRichards) [inline code/game/objects/explosion.dm]: Yes, that's good, but i you look where this definition is, you're defining a turf as a turf.
```var/turf/Trajectory = var/turf/T```

File: code/game/objects/explosion.dm (hunk 2 above)
Comment (DeityLink) [inline code/game/objects/explosion.dm]: Already replace that completely. I appreciate that you want to review my code but please give me 5 minutes to commit the latest changes.

Comment (RemieRichards) [inline code/game/objects/explosion.dm]: Yeah that's cool, I just dig into code the moment I see it.

**Comment by DeityLink** [comment]
Alright, I think I've tied all the loose ends now, here, have a gif.

explos

**Comment by DeityLink** [comment]
The main differences with tg's old explosion_recursive beside the cheaper approach and smaller number of code lines, are that the obstacles are still fully affected by the blast (only the tiles behind them are protected), and that since the protections are checked before each turf is ex_act 'd, a turf that could be theoretically be protected might end up still being affected if its protection has been destroyed by the time the explosion reaches it, WHICH in my honest opinion offers a more realistic portrayal of explosions (which happen over a few seconds in real life, rather than at a singular instant)

**Comment by RemieRichards** [comment]
@DeityLink for your debug verb, have the Maptext of the turf also be either: "Dev","Heavy" or "Light", it all fits and adds to the effect, meaning less colour memorisation.

**Comment by RemieRichards** [comment]
Also I noticed (hilariously) if you give something a negative ```explosion_block``` it improves the explosion.

**Comment by RemieRichards** [comment]
https://github.com/tgstation/-tg-station/pull/9412, for your own tracking needs.

**Comment by DeityLink** [comment]
I assume that people can memorize 3 colours, and they're pretty explicit too.

And yes, since explosion_block is added to the "distance from the epicenter", if you give a negative one you'll end up "closer" to the epicenter, thus the turf and the atoms on it will take more damage.
It won't however increase the max range of the explosion itself.

**Comment by FudgePucker** [comment]
I think that standing behind a wall when a bomb goes off should deal more damage to people.
The wall doesn't vanish realistically, it turns into tiny metal fragments going faster than the speed of sound, right into your face, ripping off your face, shattering every bone in your body.

**Comment by DeityLink** [comment]
You're invited to code shrapnel if you wish, and even the surgery steps for doctors to remove it from patients.

**Comment by d3athrow** [comment]
Easy enough just get the direction of the epicenter to the walland then check shit adjacent to it on that side in ex_act

**Comment by FudgePucker** [comment]
Oh man shrapnel would be fun to implement in the game.
Anti personnel grenades that actually do more damage to people than the environment.
Graytide IEDs put into lockers that blow off the limbs of shitcurity.
Pulse rifle killing everything in the room you are breaking into before you even get in there.

File: code/game/turfs/simulated/floor_types.dm [hunk 3]
```diff
@@ -121,6 +121,26 @@
...
 
+/turf/simulated/floor/engine/ex_act(severity)
+	switch(severity)
+		if(1.0)
+			if(prob(80))
+				src.ReplaceWithLattice()
+			else if(prob(50))
+				src.ChangeTurf(under_turf)
+			else
+				var/turf/simulated/floor/F = src
```
Comment (ComicIronic) [inline code/game/turfs/simulated/floor_types.dm]: Why are you casting src?

Comment (DeityLink) [inline code/game/turfs/simulated/floor_types.dm]: There probably was a reason when I first wrote that but I can't see it anymore myself.

**Pull request merged**